### ISSUE TITLE
Static properties on components break tree-shaking

### ISSUE THREAD

**Issue by mbrowne** [styled-components/babel-plugin-styled-components#245]
In our app we are finding that any time a component that uses styled-components has a static property added to it, that component is always included in the bundle even if it's not imported, i.e. tree shaking does not work. For example:

```js
const Wrapper = styled.div`
    color: blue;
`

function MyComponent() {
    return <Wrapper>hi</Wrapper>
}
// this breaks tree-shaking
MyComponent.displayName = 'Foo'
```

In some cases we're able to remove the `displayName` or use default function argument values instead of `defaultProps`, but there are other cases where we do need to use these properties.

This seems to be caused by the way terser (or uglify-js) removes dead code and isn't really an issue in babel-plugin-styled-components per se, but it would be great if babel-plugin-styled-components could include an option to address this.

**Comment by mbrowne** [comment]
UPDATE: This only seems to affect class components. Some of our function components were failing to tree-shake as well, but it looks like that was actually because they were being used inside of class components that were affected by this issue.

I also discovered that `React.forwardRef()` breaking tree-shaking, but that's unrelated to styled-components.

**Comment by mbrowne** [comment]
(Note: deleted my previous comment. I got confused in my testing and what I wrote was incorrect. This is an issue for both class components and function components.)

**Comment by mxstbr** [comment]
This is very interesting, thank you for digging into this! :pray: 

A PR that makes adds the IIFE wrapper (next to the PURE comment) when setting `pure: true` would be much appreciated, I definitely think the Babel plugin should be doing whatever it can to make tree shaking possible.

**Comment by mbrowne** [comment]
Great, glad you are open to including this. I can probably work on a PR for this sometime in the next week or two.

**Comment by mxstbr** [comment]
Amazing, can't wait @mbrowne! :pray:

**Comment by mbrowne** [comment]
For anyone who's interested, I created a test repo so that others can easily reproduce this issue, and also so that we have something to test against to ensure that my PR (when it's ready) covers all the bases:
https://github.com/mbrowne/CRA-error-template/tree/styled-components-tree-shaking
(be sure to check out this specific branch)

If you run `yarn build` on my demo, you can see that only `Component1` tree-shakes correctly. (Earlier I thought that even that would break, because it seemed like this issue was affecting all of our library components, but fortunately in that simple case it's already working correctly.)

**Comment by ammmze** [comment]
Is this issue actually specific to styled-components? Maybe i've got something goofed in my project i'm in currently, but regardless of using styled-components, if I add a static property (`MyComponent.displayName = 'MyComponent'`) to the function component, then it no longer gets removed during dead code removal. But if I remove the static property then it gets removed. This was also the case prior to me adding `babel-plugin-styled-comonents` to my project.

**Comment by mbrowne** [comment]
@ammmze It isn't only an issue for styled-components. But in order to address it in a different plugin, that plugin would have to be aware of how to detect styled-components, and duplicate all that detection logic that's already included in `babel-plugin-styled-components`. Also, `babel-plugin-styled-components` already adds some PURE annotations in order to make tree-shaking work correctly, so it's logical that it should also handle the static properties case (`displayName`, etc).

**Comment by ammmze** [comment]
Gotcha! So it looks like with the fix you have in the PR, it only seems to take effect for components are literally made from styled components (i.e. ``export const Foo = styled.div`color: blue`;``) or functions that return styled components. Do you have any tips for function components that just return native react elements? Something like below doesn't seem to get wrapped like styled-components and functions that return styled-components.

```javascript
const Foo = (props) => (<div {...props}>Foo</div>);
Foo.displayName = 'Foo';
export default Foo;
```

**Comment by mbrowne** [comment]
@ammmze Yeah, I had forgotten about that until someone pointed it out to me recently. They told me they're now using both my PR to this repo and my older plugin, [babel-plugin-pure-static-props](https://github.com/mbrowne/babel-plugin-pure-static-props). I still think that all the use cases related to styled-components should be handled by babel-plugin-styled-components, and that other plugins shouldn't be concerning themselves with detecting styled components. But for other use cases, I'm thinking maybe I should un-deprecate babel-plugin-pure-static-props and update it with the efficiency improvements I made as part of [this PR](https://github.com/styled-components/babel-plugin-styled-components/pull/248).

**Comment by ammmze** [comment]
👍🏻  That is what I'm thinking. I think it makes perfect sense for the `babel-plugin-styled-components` to apply it to any component is created from styled components (i.e. ``const Foo = styled.div`color: blue`;``), and functions that return styled-components...anything related to styled-components. But for vanilla I think it should be a separate plugin. I did try adding your deprecated plugin, but ran into some issues.

It blew up when it hit an old class component we have in our stuff...

**Comment by ammmze** [comment]
@mbrowne FYI i've forked your project and i'm setting up test fixtures to cover some various use cases and I'll submit a PR when complete
